Bottom: 450417ea4e8a6f8c1cc0335ba306c999f803522e
Top:    87298eb6752fcca9cef63384a83b77a361b58cbc
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2012-12-22 22:29:59 +0000

Refresh of major-re-write-to-drawing-rout

---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index a139001..573edbe 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -975,16 +975,16 @@ EMark_callback (const BoxType * b, void *cl)
 }
 
 static void
-set_object_color (AnyObjectType *obj,
-                  char *warn_color, char *selected_color,
-                  char *found_color, char *normal_color)
+set_object_color (AnyObjectType *obj, char *warn_color, char *selected_color,
+                  char *connected_color, char *found_color, char *normal_color)
 {
   char *color;
 
-  if      (warn_color     != NULL && TEST_FLAG (WARNFLAG,     obj)) color = warn_color;
-  else if (selected_color != NULL && TEST_FLAG (SELECTEDFLAG, obj)) color = selected_color;
-  else if (found_color    != NULL && TEST_FLAG (FOUNDFLAG,    obj)) color = found_color;
-  else                                                              color = normal_color;
+  if      (warn_color      != NULL && TEST_FLAG (WARNFLAG,      obj)) color = warn_color;
+  else if (selected_color  != NULL && TEST_FLAG (SELECTEDFLAG,  obj)) color = selected_color;
+  else if (connected_color != NULL && TEST_FLAG (CONNECTEDFLAG, obj)) color = connected_color;
+  else if (found_color     != NULL && TEST_FLAG (FOUNDFLAG,     obj)) color = found_color;
+  else                                                                color = normal_color;
 
   gui->graphics->set_color (Output.fgGC, color);
 }
@@ -992,18 +992,7 @@ set_object_color (AnyObjectType *obj,
 static void
 set_layer_object_color (LayerType *layer, AnyObjectType *obj)
 {
-  set_object_color (obj, NULL, layer->SelectedColor, PCB->ConnectedColor, layer->Color);
-}
-
-static void
-set_pv_color (PinType *pv, int type)
-{
-  if (TEST_FLAG (WARNFLAG, pv))          gui->graphics->set_color (Output.fgGC, PCB->WarnColor);
-  else if (TEST_FLAG (SELECTEDFLAG, pv)) gui->graphics->set_color (Output.fgGC, (type == VIA_TYPE) ? PCB->ViaSelectedColor
-                                                                                                   : PCB->PinSelectedColor);
-  else if (TEST_FLAG (FOUNDFLAG, pv))    gui->graphics->set_color (Output.fgGC, PCB->ConnectedColor);
-  else                                   gui->graphics->set_color (Output.fgGC, (type == VIA_TYPE) ? PCB->ViaColor :
-                                                                                                     PCB->PinColor);
+  set_object_color (obj, NULL, layer->SelectedColor, PCB->ConnectedColor, PCB->FoundColor, layer->Color);
 }
 
 static void
@@ -1079,7 +1068,9 @@ _draw_pv (PinType *pv, bool draw_hole)
 static void
 draw_pin (PinType *pin, bool draw_hole)
 {
-  set_pv_color (pin, PIN_TYPE);
+  set_object_color ((AnyObjectType *) pin, PCB->WarnColor, PCB->PinSelectedColor,
+                    PCB->ConnectedColor, PCB->FoundColor, PCB->PinColor);
+
   _draw_pv (pin, draw_hole);
 }
 
@@ -1101,7 +1092,9 @@ pin_inlayer_callback (const BoxType * b, void *cl)
 static void
 draw_via (PinType *via, bool draw_hole)
 {
-  set_pv_color (via, VIA_TYPE);
+  set_object_color ((AnyObjectType *) via, PCB->WarnColor, PCB->ViaSelectedColor,
+                    PCB->ConnectedColor, PCB->FoundColor, PCB->ViaColor);
+
   _draw_pv (via, draw_hole);
 }
 
@@ -1179,7 +1172,7 @@ static void
 draw_pad (PadType *pad)
 {
   set_object_color ((AnyObjectType *)pad, PCB->WarnColor,
-                    PCB->PinSelectedColor, PCB->ConnectedColor,
+                    PCB->PinSelectedColor, PCB->ConnectedColor, PCB->FoundColor,
                     FRONT (pad) ? PCB->PinColor : PCB->InvisibleObjectsColor);
 
   _draw_pad (Output.fgGC, pad, false, false);
@@ -1227,7 +1220,7 @@ hole_callback (const BoxType * b, void *cl)
   if (TEST_FLAG (HOLEFLAG, pv))
     {
       set_object_color ((AnyObjectType *) pv, PCB->WarnColor,
-                        PCB->ViaSelectedColor, NULL, Settings.BlackColor);
+                        PCB->ViaSelectedColor, NULL, NULL, Settings.BlackColor);
 
       gui->graphics->set_line_cap (Output.fgGC, Round_Cap);
       gui->graphics->set_line_width (Output.fgGC, 0);
